app_flutter/lib/widgets/task_grid.dart

```diff
@@ -47,12 +49,367 @@ class TaskGridContainer extends StatelessWidget {
+      'showCirrus': showCirrus.toString(),
+    if (showLuci != defaultFilter.showLuci)
+      'showLuci': showLuci.toString(),
+  };
+
+  String _stringFor(String label, Object value) => value == null ? '' : '$label: $value,';
+
+  @override
+  String toString() {
+    return 'TaskGridFilter('
+        '${_stringFor('taskFilter',    taskFilter?.pattern)}'
+        '${_stringFor('authorFilter',  authorFilter?.pattern)}'
+        '${_stringFor('messageFilter', messageFilter?.pattern)}'
+        '${_stringFor('showAndroid',   showAndroid)}'
+        '${_stringFor('showIos',       showIos)}'
+        '${_stringFor('showWindows',   showWindows)}'
+        '${_stringFor('showCirrus',    showCirrus)}'
+        '${_stringFor('showLuci',      showLuci)}'
+        ')';
+  }
+
+  @override
+  int get hashCode {
+    return hashValues(
+      taskFilter,
+      authorFilter,
+      messageFilter,
+      showAndroid,
+      showIos,
+      showWindows,
+      showCirrus,
+      showLuci,
+    );
+  }
+
+  @override
+  bool operator ==(Object other) {
+    if (other.runtimeType != runtimeType) {
+      return false;
+    }
+    return other is TaskGridFilter &&
+        taskFilter == other.taskFilter &&
+        authorFilter == other.authorFilter &&
+        messageFilter == other.messageFilter &&
+        showAndroid == other.showAndroid &&
+        showIos == other.showIos &&
+        showWindows == other.showWindows &&
+        showCirrus == other.showCirrus &&
+        showLuci == other.showLuci
+    ;
+  }
+}
+
+class TaskGridFilterWidget extends StatefulWidget {
+  const TaskGridFilterWidget(this.filterNotifier, this.onClose);
+
+  final ValueNotifier<TaskGridFilter> filterNotifier;
+  final Function() onClose;
+
+  @override
+  State createState() => TaskGridFilterState();
+}
+
+class TaskGridFilterState extends State<TaskGridFilterWidget> {
+  @override
+  void initState() {
+    super.initState();
+    widget.filterNotifier.addListener(_update);
+    taskController = TextEditingController(text: filter.taskFilter?.pattern ?? '');
+    authorController = TextEditingController(text: filter.authorFilter?.pattern ?? '');
+    messageController = TextEditingController(text: filter.messageFilter?.pattern ?? '');
+  }
+
+  void _update() {
+    setState(() {});
+  }
+
+  @override
+  void dispose() {
+    widget.filterNotifier.removeListener(_update);
+    super.dispose();
+  }
+
+  static TextStyle labelStyle = const TextStyle(
+    color: Colors.black,
+    fontSize: 16.0,
+    fontWeight: FontWeight.normal,
+    fontStyle: FontStyle.normal,
+    decoration: TextDecoration.none,
+  );
+
+  TaskGridFilter get filter => widget.filterNotifier.value;
+  TextEditingController taskController;
+  TextEditingController authorController;
+  TextEditingController messageController;
+
+  void _setFilter(TaskGridFilter newFilter) {
+    widget.filterNotifier.value = newFilter;
+  }
+
+  void _newTaskFilter(RegExp newRegExp) {
```

Comment:
This seems verbose. Can these new functions be condensed into a single function?

Reply:
I was thinking about that. I'll see what I can do.

I was also thinking about making a generic filter object with arbitrary filters so we could do the same thing with other dashboards.

(FWIW, I have an alternate implementation of the benchmark dashboard in the [benchmark utilities repo](https://github.com/flar/flutter_benchmark_utils/blob/master/README.md) that also has filtering...)

Reply:
It's great to see the benchmark utilities repo! If you want to have it replace the AngularDart dashboard let me know. It's trivial to setup a new page that has all the data needed, I just never got around to building graph widgets that could show the data.

Reply:
I solved this one by re-using the map constructor as a copyWithMap() method so I don't need a per-attribute updater.

Reply:
I've broken this out into a FilterPropertySheet widget that can layout and edit arbitrary filtering properties and the TaskGridFilter now implements that "interface". A lot of the redundancy here went away as well at the expense of some generic field get/set functions. I might revisit that in the future to see if I can get it simplified some more.